immer/detail/type_traits.hpp

```diff
@@ -0,0 +1,193 @@
+// Copyright (C) 2016, 2017 Juan Pedro Bolivar Puente
+//
+// This file is part of immer.
+//
+// immer is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// immer is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with immer.  If not, see <http://www.gnu.org/licenses/>.
+//
+
+#pragma once
+
+#include <algorithm>
+#include <iterator>
+#include <memory>
+#include <type_traits>
+#include <utility>
+
+namespace immer {
+namespace detail {
+
+template <typename... Ts>
+struct make_void { using type = void; };
+
+template <typename... Ts>
+using void_t = typename make_void<Ts...>::type;
+
+template <typename T, typename = void>
+struct is_dereferenceable : std::false_type {};
+
+template <typename T>
+struct is_dereferenceable<T, void_t<decltype(*(std::declval<T&>()))>> :
+    std::true_type {};
+
+template <typename T>
+constexpr bool is_dereferenceable_v = is_dereferenceable<T>::value;
+
+template<typename T, typename U=T, typename = void>
+struct is_equality_comparable : std::false_type {};
+
+template<typename T, typename U>
+struct is_equality_comparable
+<T, U, std::enable_if_t
+ <std::is_same
+  <bool, decltype(std::declval<T&>() == std::declval<U&>())>::value>> :
+    std::true_type {};
+
+template<typename T, typename U=T>
+constexpr bool is_equality_comparable_v = is_equality_comparable<T, U>::value;
+
+template<typename T, typename U=T, typename = void>
+struct is_inequality_comparable : std::false_type {};
+
+template<typename T, typename U>
+struct is_inequality_comparable
+<T, U, std::enable_if_t
+ <std::is_same
+  <bool, decltype(std::declval<T&>() != std::declval<U&>())>::value>> :
+    std::true_type {};
+
+template<typename T, typename U=T>
+constexpr bool is_inequality_comparable_v =
+  is_inequality_comparable<T, U>::value;
+
+template <typename T, typename = void>
+struct is_preincrementable : std::false_type {};
+
+template <typename T>
+struct is_preincrementable
+<T, std::enable_if_t
+ <std::is_same<T&, decltype(++(std::declval<T&>()))>::value>> :
+    std::true_type {};
+
+template <typename T>
+constexpr bool is_preincrementable_v = is_preincrementable<T>::value;
+
+template <typename T, typename U=T, typename = void>
+struct is_subtractable : std::false_type {};
+
+template <typename T, typename U>
+struct is_subtractable
+<T, U, void_t<decltype(std::declval<T&>() - std::declval<U&>())>> :
+    std::true_type {};
+
+template <typename T, typename U = T>
+constexpr bool is_subtractable_v = is_subtractable<T, U>::value;
+
+using std::swap;
```

Comment:
I just noticed this. Could you move this line (and the implementation of `is_swapable_with`) to a nested namespace `is_swapable_with_impl` so `std::swap` is not leaked into the `immer::detail` namespace?